src/msw/webview_edge.cpp

```diff
@@ -0,0 +1,777 @@
+PFNWXCreateWebView2EnvironmentWithDetails wxCreateWebView2EnvironmentWithDetails = NULL;
+PFNWXGetWebView2BrowserVersionInfo wxGetWebView2BrowserVersionInfo = NULL;
+
+wxDynamicLibrary wxWebViewEdge::ms_loaderDll;
+
+bool wxWebViewEdge::IsAvailable()
+{
+    if (ms_isAvailable == -1)
+    {
+        if (!Initialize())
+            ms_isAvailable = 0;
+        else
+            ms_isAvailable = 1;
+    }
+
+    return (ms_isAvailable == 1);
+}
+
+bool wxWebViewEdge::Initialize()
+{
+#define RESOLVE_LOADER_FUNCTION(type, funcname)                        \
+    wx##funcname = (type)ms_loaderDll.GetSymbol(wxT(#funcname));       \
+    if ( !wx##funcname )                                               \
+        return false
+
+    // WebView2 is only available for Windows 7 or newer
+    if (!wxCheckOsVersion(6, 1))
+        return false;
+
+    if (!ms_loaderDll.Load("WebView2Loader.dll", wxDL_DEFAULT | wxDL_QUIET))
+        return false;
+
+    // Try to load functions from loader DLL
+    RESOLVE_LOADER_FUNCTION(PFNWXCreateWebView2EnvironmentWithDetails, CreateWebView2EnvironmentWithDetails);
+    RESOLVE_LOADER_FUNCTION(PFNWXGetWebView2BrowserVersionInfo, GetWebView2BrowserVersionInfo);
+
+    // Check if a Edge browser can be found by the loader DLL
+    LPWSTR versionStr;
+    if (SUCCEEDED(wxGetWebView2BrowserVersionInfo(NULL, &versionStr)))
+    {
+        if (versionStr)
+            return true;
+    }
+
+    return false;
+}
+
+void wxWebViewEdge::Uninitalize()
+{
+    if (ms_isAvailable == 1)
+    {
+        ms_loaderDll.Unload();
+        ms_isAvailable = -1;
+    }
+}
+
+wxWebViewEdge::~wxWebViewEdge()
+{
+    if (m_webView)
+    {
+        // TOOD: Remove additional events
+        m_webView->remove_NavigationCompleted(m_navigationCompletedToken);
+        m_webView->remove_NavigationStarting(m_navigationStartingToken);
+        m_webView->remove_DocumentStateChanged(m_documentStateChangedToken);
+        m_webView->remove_NewWindowRequested(m_newWindowRequestedToken);
+    }
+}
+
+bool wxWebViewEdge::Create(wxWindow* parent,
+    wxWindowID id,
+    const wxString& url,
+    const wxPoint& pos,
+    const wxSize& size,
+    long style,
+    const wxString& name)
+{
+    m_initialized = false;
+    m_isBusy = false;
+
+    m_historyLoadingFromList = false;
+    m_historyEnabled = true;
+    m_historyPosition = -1;
+
+    if (!wxControl::Create(parent, id, pos, size, style,
+        wxDefaultValidator, name))
+    {
+        return false;
+    }
+
+    if (!IsAvailable())
+        return false;
+
+    m_pendingURL = url;
+
+    Bind(wxEVT_SIZE, &wxWebViewEdge::OnSize, this);
+
+    wxString userDataPath = wxStandardPaths::Get().GetUserLocalDataDir();
+
+    HRESULT hr = wxCreateWebView2EnvironmentWithDetails(
+        nullptr,
+        userDataPath.wc_str(),
+        nullptr,
+        Callback<IWebView2CreateWebView2EnvironmentCompletedHandler>(
+            [this](HRESULT WXUNUSED(result), IWebView2Environment* environment) -> HRESULT
```

Comment:
I'm not really against using C++11 in this file (as it can only be compiled with MSVS 2015 anyhow), but it's not clear to me that using lambdas improves readability of this code, especially for the longer functions below. Why not just make them normal functions/methods?

Reply:
This isn't really about lambda vs function. The `Microsoft::WRL::Callback<>` template implements an instance of event call back type (in this case `IWebView2CreateWebView2EnvironmentCompletedHandler`).
The template takes a lambda or a function pointer (which wouldn't work with a method pointer, if I'm not mistaken).
I think it improves readability immensely compared to the alternative where `wxWebViewEdge` would have to implement all these event interfaces.

From a pure technical point of view it could obviously be done, but I've no interest in doing it, especially not in this case where the underlying SDK recommends Visual Studio 2015 or newer.

Reply:
> This isn't really about lambda vs function. The `Microsoft::WRL::Callback<>` template implements an instance of event call back type (in this case `IWebView2CreateWebView2EnvironmentCompletedHandler`).
> The template takes a lambda or a function pointer (which wouldn't work with a method pointer, if I'm not mistaken).

It looks like `WRL::Callback` [can be constructed](https://docs.microsoft.com/en-us/cpp/cppcx/wrl/callback-function-wrl?view=vs-2019) from an object and its method, so I thought it would work. I could be wrong though, I've never used it. But my remark was really _just_ about lambdas vs functions.

Reply:
Didn't know that constructor (never saw it used in the samples), but with it readability can definitely be improved (especially for the longer lambdas).